app/actions/views/channel.js

```diff
@@ -156,23 +164,72 @@ export function loadProfilesAndTeamMembersForDMSidebar(teamId) {
     };
 }
 
+export function switchToChannel(channelId, displayName) {
+    return async (dispatch, getState) => {
+        const state = getState();
+        const currentTeamId = getCurrentTeamId(state);
+        const currentChannelId = getCurrentChannelId(state);
+
+        dispatch(loadPostsIfNecessaryWithRetry(channelId));
+
+        const actions = [
+            setLoadMorePostsVisible(true),
+            setChannelLoading(channelId !== currentChannelId),
+            recordLoadTime('Switch Channel', 'channelSwitch'),
+            {
+                type: ChannelTypes.SELECT_CHANNEL,
+                data: channelId,
+            },
+            {
+                type: ViewTypes.SET_LAST_CHANNEL_FOR_TEAM,
+                teamId: currentTeamId,
+                channelId,
+            },
+            {
+                type: ViewTypes.SET_INITIAL_POST_VISIBILITY,
+                data: channelId,
+            },
+        ];
+
+        if (displayName) {
+            actions.push(setChannelDisplayName(displayName));
+        }
+
+        dispatch(batchActions(actions));
+        dispatch(markChannelAsRead(channelId, currentChannelId));
+        if (channelId !== currentChannelId) {
+            dispatch(markChannelAsViewed(currentChannelId));
+        }
+    };
+}
+
 export function loadPostsIfNecessaryWithRetry(channelId) {
     return async (dispatch, getState) => {
         const state = getState();
         const {posts, postsInChannel} = state.entities.posts;
         const postsIds = postsInChannel[channelId];
-
         const time = Date.now();
 
-        let loadMorePostsVisible = true;
-        let received;
-        if (!postsIds || postsIds.length < ViewTypes.POST_VISIBILITY_CHUNK_SIZE) {
-            // Get the first page of posts if it appears we haven't gotten it yet, like the webapp
-            received = await retryGetPostsAction(getPosts(channelId), dispatch, getState);
-
+        const receivedPosts = (received) => {
```

Comment:
0/5 on this, but why move this up here? If we await on both `retryGetPostsAction` calls below instead of using `then`, we can put this at the bottom of the function